cloud_testing/kotlin_poc/src/main/kotlin/ftl/json/SavedMatrix.kt

```diff
@@ -45,12 +45,17 @@ class SavedMatrix(matrix: TestMatrix) {
     }
 
     private fun finished(matrix: TestMatrix) {
-        val toolResultsStep = matrix.firstToolResults()
-        if (toolResultsStep != null) {
-            val step = GcToolResults.getResults(toolResultsStep)
-            testDurationSeconds = step.testExecutionStep.testTiming.testProcessDuration.seconds
-            runDurationSeconds = step.runDuration.seconds
-            billableMinutes = Billing.billableMinutes(testDurationSeconds)
+        billableVirtualMinutes = 0
+        billablePhysicalMinutes = 0
+        matrix.testExecutions.forEach {
+            val step = GcToolResults.getResults(it.toolResultsStep)
+            val billableMinutes = Billing.billableMinutes(step.testExecutionStep.testTiming.testProcessDuration.seconds)
+            val androidDevice = it.environment.androidDevice
+            if (androidDevice != null && AndroidCatalog.isVirtualDevice(androidDevice.androidModelId)) {
```

Comment:
I think you could do something like this:

> AndroidCatalog.isVirtualDevice(androidDevice)

then the null check / androidModelId would move into the helper method.